src/proto/connection.rs

```diff
@@ -292,52 +289,60 @@ where
             // The order here matters:
             // - poll_go_away may buffer a graceful shutdown GOAWAY frame
             // - If it has, we've also added a PING to be sent in poll_ready
-            if let Some(reason) = try_ready!(self.poll_go_away()) {
-                if self.go_away.should_close_now() {
-                    if self.go_away.is_user_initiated() {
-                        // A user initiated abrupt shutdown shouldn't return
-                        // the same error back to the user.
-                        return Ok(Async::Ready(()));
-                    } else {
-                        return Err(RecvError::Connection(reason));
+            match ready!(self.poll_go_away(cx)) {
+                Some(Ok(reason)) => {
+                    if self.go_away.should_close_now() {
+                        if self.go_away.is_user_initiated() {
+                            // A user initiated abrupt shutdown shouldn't return
+                            // the same error back to the user.
+                            return Poll::Ready(Ok(()));
+                        } else {
+                            return Poll::Ready(Err(RecvError::Connection(reason)));
+                        }
                     }
+                    // Only NO_ERROR should be waiting for idle
+                    debug_assert_eq!(
+                        reason,
+                        Reason::NO_ERROR,
+                        "graceful GOAWAY should be NO_ERROR"
+                    );
                 }
-                // Only NO_ERROR should be waiting for idle
-                debug_assert_eq!(reason, Reason::NO_ERROR, "graceful GOAWAY should be NO_ERROR");
+                Some(Err(e)) => return Poll::Ready(Err(e.into())),
+                None => (),
             }
-            try_ready!(self.poll_ready());
+            ready!(self.poll_ready(cx))?;
 
-            match try_ready!(self.codec.poll()) {
-                Some(Headers(frame)) => {
+            match ready!(Pin::new(&mut self.codec).poll_next(cx)) {
```

Comment:
I thought there was a `Try` impl for `Poll<Option<Result<T, E>>` so all the `Ok(..)` patterns could be skipped. If not, oh well.

Reply:
Done